### ISSUE TITLE
Crash when creating a .subscriber socket

### ISSUE THREAD

**Issue by erparker** [azawawi/SwiftyZeroMQ#13]
I am following the pub-sub example listed, modified slightly because I didn't need the publisher (running the unedited example gives me the same error), and I can't get past the creation of a subscriber without crashing. the code looks like the following:

    public var endpoint = "tcp://127.0.0.1:5550"
    
    public func subscribe(topic: String, callback: @escaping (_ name: String,_ recieved: String) -> Void) {
        do {
            let context = try SwiftyZeroMQ.Context()
            let subscriber = try context.socket(.subscribe)
            
            let subscribers = [
                subscriber: "Subscriber #1"
            ]
            
            try subscriber.setSubscribe(topic)

            // Brief wait to let everything hook up
            usleep(250)
            
            let poller = SwiftyZeroMQ.Poller()
            try poller.register(socket: subscriber, flags: .pollIn)
            
            func pollAndRecv() throws {
                let socks = try poller.poll(timeout: 1000)
                for subscriber in socks.keys {
                    let name = subscribers[subscriber]
                    if socks[subscriber] == SwiftyZeroMQ.PollFlags.pollIn {
                        let text = try subscriber.recv(options: .dontWait)
                        callback("\(name)", "\(text)")
                        print("\(name): received '\(text)'")
                    } else {
                        print("\(name): Nothing")
                    }
                }
                print("---")
            }
            
            try pollAndRecv();
            
        } catch {
            print(error)
        }
    }

The stack looks like this: 
![screen shot 2017-01-18 at 4 22 53 pm](https://cloud.githubusercontent.com/assets/5798942/22087517/931700ee-dd9b-11e6-862d-7069b57bc311.png)

I don't get any error messages in the console, execution simply stops on 
"dyld`dyld_fatal_error:
->  0x120009088 <+0>: brk    #0x3"

and the last thing called in code was this 
```
public init(context: Context, type : SocketType) throws {
    // Create socket
    let p : UnsafeMutableRawPointer? = zmq_socket(context.handle,
        type.rawValue)
    guard p != nil else {
        throw ZeroMQError.last
    }

    // Now we can assign socket handle safely
    handle = p!
    cleanupNeeded = true
}
```

with the debugger failing on this line: 
let p : UnsafeMutableRawPointer? = zmq_socket(context.handle,
        type.rawValue)

Any ideas or pointers would be much appreciated! Thanks for all the great work

**Comment by azawawi** [comment]
@erparker Thanks for reporting this interesting bug.

- Which platform and simulator (if any) are you targeting? 
- Is `subscribe` called more than once?

**Comment by erparker** [comment]
I am building a debug scheme to an iPad mini 2 running iOS 10.1. Subscribe is not called more than once in my code.

One thing to note, I have a shared project that contains a number of targets, each a framework that gets used in a number of applications. We do this as a way to have centralized models, networking code, and UI that can easily be shared. I added the zmq framework to this shared project, which is how I am gaining access to the zmq code. There aren't any compiler warnings or runtime errors specifically calling out a bad configuration, nor do I see any linker errors surrounding finding the zmq framework. But I thought I would point this out in case it could affect the running of this code some how. 

We're also using New Relic, and I don't know how they implemented their framework, just wondering if there's a chance the socket layer isn't playing nicely.

**Comment by azawawi** [comment]
Cool then try it without other frameworks that affect the socket layer and see if it works. But still it failed in `clock_t` from which I patched earlier for it to work on iOS if I remember correctly. Please see https://github.com/azawawi/libzmq-ios/blob/master/libzmq.rb#L262. Is it failing on the simulator or the actual device?

**Comment by erparker** [comment]
Sorry, didn't make that clear, this is an actual device that it's failing on. Calling subscribe once inside AppDelegate.
I'll turn off all frameworks next chance I get and see how that affects this. Thanks!

**Comment by azawawi** [comment]
No problem. This is the culprit of the [clock_t crash](https://github.com/zeromq/libzmq/blob/v4.2.1/src/clock.cpp#L120) so far. I am going to try it now on my iPhone and see if it exhibits the same crash. Sometimes the simulators and travis CI give you a false sense of security :).

**Comment by azawawi** [comment]
I just ran it on my old iPhone 5 (iOS 10.2) using a USB cable and it worked with and without a publisher. Printed ```Subscriber #1: Nothing```. iOS compatibility was set to 9.

**Comment by erparker** [comment]
So I just ran this exact code, no modifications to potentially competing frameworks, and it worked on an iPad Mini 2 retina (model A1489) running 10.1.1, which is puzzling because the other device that is having the crash is the exact same model number A1489. I actually haven't confirmed if they're both running the same version of iOS, but I will need to look tomorrow since I don't have access to the original device today.

Also, for what it's worth, I ran this code on my iPhone 6s running 10.1.1 is everything worked fine.

I'm going to run this on as many devices as possible and see if I can replicate that strange crash condition.

**Comment by erparker** [comment]
quick update: I've tested two iPads, one iOS 9.3.5, the other 9.3.2, and they both have the same clock_t crash. Devices running iOS 10+ don't crash. (i was incorrect in my initial report that the crashing device was running iOS 10)

**Comment by azawawi** [comment]
So basically this bug "[ZeroMQ 4.1.5 runtime crash on iOS 10 SDK](https://lists.zeromq.org/pipermail/zeromq-dev/2016-October/031022.html)" that I reported earlier in 4.1 is hitting us in 4.2. I think they changed the implementation in 4.2. We need to get the workaround patch again.

**Comment by azawawi** [comment]
We need to revert the following pull https://github.com/zeromq/libzmq/pull/2176 and rebuild using [libzmq.rb](https://github.com/azawawi/libzmq-ios/blob/master/libzmq.rb). I am currently away from my laptop for a couple of days (sick with flu). I will do it once I am back.

**Comment by azawawi** [comment]
@erparker Please try https://github.com/azawawi/SwiftyZeroMQ/commit/db0051659a44db7f56c46b7755af32d4a2b8fce0. Sorry for the delay.

**Comment by azawawi** [comment]
@erparker Seems that crash is solved according to #14 with the latest commit. Please let me know the test results from your side if possible.

**Comment by azawawi** [comment]
I am closing this one. Please reopen if needed. Thanks for your time.
